### PR TITLE
Us31 33

### PR DESCRIPTION
**Pull request by perryr16**
31: pets with approved apps (adoptable = pending) cannot be deleted
32: deleting a pet removes it form favorites
33: flash message when you update/create/incomplete a pet

### CODE REVIEW

File: app/controllers/pets_controller.rb [hunk 1]
```diff
@@ -12,11 +12,18 @@ def new
...
     shelter = Shelter.find(params[:id])
-    pet = shelter.pets.create!(pet_params)
+    pet = shelter.pets.create(pet_params)
 
-    redirect_to "/shelters/#{shelter.id}/pets"
+    if pet.save
+      flash[:message] = "#{pet.name} has been created!"
+      redirect_to "/shelters/#{shelter.id}/pets"
+    else
+      flash[:message] = "You must fill out #{empty_params}"
```
Comment (perryr16) [inline app/controllers/pets_controller.rb:21]: this message now informs users in exactly which params have not been entered

File: app/controllers/pets_controller.rb [hunk 2]
```diff
@@ -28,25 +35,31 @@ def edit
...
+    pet.update(pet_params)
 
     redirect_to '/pets'
   end
 
   def destroy
+    pet_id = params[:id]
+
+    if !session[:favorite].nil?
+      session[:favorite].delete(pet_id.to_s)
```
Comment (perryr16) [inline app/controllers/pets_controller.rb:47]: if pet is deleted, it is removed from the session

File: app/controllers/pets_controller.rb [hunk 3]
```diff
@@ -28,25 +35,31 @@ def edit
...
+    end
+
     Pet.destroy(params[:id])
     redirect_to '/pets'
   end
 
-  def is_adoptable
+  def adoptable
```
Comment (perryr16) [inline app/controllers/pets_controller.rb:54]: was told "underscores" are special characters

File: app/controllers/pets_controller.rb [hunk 4]
```diff
@@ -65,5 +78,13 @@ def pet_params
                   :adoptable)
   end
 
+  def empty_params
+    empty_params = []
+    empty_params << "Name" if params[:name] == ""
+    empty_params << "Age" if params[:age] == ""
+    empty_params << "Image" if params[:image] == ""
+    empty_params.join(", ")
+  end
```
Comment (perryr16) [inline app/controllers/pets_controller.rb:87]: if any of the validated fields are not entered, add to a list to display in a flash message

File: app/controllers/shelters_controller.rb [hunk 5]
```diff
@@ -58,4 +58,16 @@ def shelter_params
...
 
+  def empty_params
+    empty_params = []
+    empty_params << "Name" if params[:name] == ""
+    empty_params << "Address" if params[:address] == ""
+    empty_params << "City" if params[:city] == ""
+    empty_params << "State" if params[:state] == ""
+    empty_params << "Zip" if params[:zip] == ""
+    empty_params.join(", ")
+  end
```
Comment (perryr16) [inline app/controllers/shelters_controller.rb:69]: if any of the validated fields are not entered, add to a list to display in a flash message

File: app/views/pets/index.html.erb [hunk 6]
```diff
@@ -37,11 +39,16 @@
       <a class="edit" href="/pets/<%= pet.id %>/edit">Update Pet</a><br><br>
 
       <section class="delete-form">
-        <%= button_to "Delete #{pet.name}", "/pets/#{pet.id}", method: :delete %>
+        <% if !pet.adoptable %>
+          <p>Cannot delete pet with approved applications</p>
```
Comment (perryr16) [inline app/views/pets/index.html.erb:43]: approved applications change a pets adoption status to `false`

File: spec/features/pets/12user_can_delete_pet_spec.rb [hunk 7]
```diff
@@ -77,4 +77,160 @@
...
+    expect(page).to have_css(".favorite_indicator", text: 2)
+
+    expect(current_path).to eq("/pets")
+    within("#pet-#{pet111x.id}")do
+      within(".delete-form")do
+        click_button("Delete P111x")
+      end
+    end
+    pet_ids = Pet.all.map{|pet| pet[:id]}
+    expect(pet_ids.include?(pet100x.id)).to eq(true)
```
Comment (perryr16) [inline spec/features/pets/12user_can_delete_pet_spec.rb:222]: this is a way I found to check if a pet has been deleted. Open to suggestions for improvement

File: app/controllers/pets_controller.rb (hunk 1 above)
Comment (muydanny) [reply app/controllers/pets_controller.rb:21]: nice!

File: app/controllers/pets_controller.rb (hunk 4 above)
Comment (muydanny) [reply app/controllers/pets_controller.rb:87]: theres that join again

**Pull request merged**